lib/web/sessions.go

```diff
@@ -282,55 +280,107 @@ func (c *SessionContext) GetAgent() (agent.Agent, *ssh.Certificate, error) {
-// associated with this session.
+// associated with this context's session.
 func (c *SessionContext) GetCertificates() (*ssh.Certificate, *x509.Certificate, error) {
-	pub, _, _, _, err := ssh.ParseAuthorizedKey(c.sess.GetPub())
+	pub, _, _, _, err := ssh.ParseAuthorizedKey(c.session.GetPub())
 	if err != nil {
 		return nil, nil, trace.Wrap(err)
 	}
-	sshcert, ok := pub.(*ssh.Certificate)
+	sshCert, ok := pub.(*ssh.Certificate)
 	if !ok {
 		return nil, nil, trace.BadParameter("not certificate")
 	}
-	tlscert, err := tlsca.ParseCertificatePEM(c.sess.GetTLSCert())
+	tlsCert, err := tlsca.ParseCertificatePEM(c.session.GetTLSCert())
 	if err != nil {
 		return nil, nil, trace.Wrap(err)
 	}
+	return sshCert, tlsCert, nil
 
-	return sshcert, tlscert, nil
 }
 
-// Close cleans up connections associated with requests
+// GetSessionID returns the ID of the underlying user web session.
+func (c *SessionContext) GetSessionID() string {
+	return c.session.GetName()
+}
+
+// Close cleans up resources associated with this context and removes it
+// from the user context
 func (c *SessionContext) Close() error {
-	closers := c.TransferClosers()
-	for _, closer := range closers {
-		c.Debugf("Closing %v.", closer)
-		closer.Close()
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	var errors []error
+	for _, clt := range c.remoteClt {
+		if err := clt.Close(); err != nil {
+			errors = append(errors, err)
+		}
 	}
-	if c.clt != nil {
-		return trace.Wrap(c.clt.Close())
+	if err := c.clt.Close(); err != nil {
+		errors = append(errors, err)
 	}
-	return nil
+	return trace.NewAggregate(errors...)
+}
+
+// getToken returns the bearer token associated with the underlying
+// session. Note that sessions are separate from bearer tokens and this
+// is only useful immediately after a session has been created to query
+// the token.
+func (c *SessionContext) getToken() types.WebToken {
+	return types.NewWebToken(types.WebTokenSpecV3{
+		Token:   c.session.GetBearerToken(),
+		Expires: c.session.GetBearerTokenExpiryTime(),
+	})
+}
+
+// expired returns whether this context has expired.
+// The context is considered expired when its bearer token TTL
+// is in the past
+func (c *SessionContext) expired(ctx context.Context) bool {
+	_, err := c.parent.readSession(ctx, proto.GetWebSessionRequest{
+		User:      c.user,
+		SessionID: c.session.GetName(),
+	})
+	if err == nil {
+		return false
+	}
+	c.log.WithError(err).Warn("Failed to query web session.")
+	expiry := c.session.Expiry()
+	if expiry.IsZero() {
+		return false
+	}
+	// Give the session some time to linger so existing users of the context
+	// have successfully disposed of them.
+	// If we remove the session immediately, a stale copy might still use the
+	// cached site clients.
+	// This is a cheaper way to avoid race without introducing object
+	// reference counters.
+	// TODO(dmitri): this leaks the implementation detail about the session TTL
+	return c.parent.clock.Since(expiry) > 2*auth.BearerTokenTTL
```

Comment:
Maybe I'm reading this incorrectly, but doesn't this basically extend the session TTL by 3 times? It allows the session to linger for 20 minutes after expiration. Is this not going to be a security concern?

Reply:
This test is only for the proxy's internal memory cache - it does not affect expiration of the actual sessions in the backend. As given in the comment, it is the attempt to provide cheap protection against a client accessing served SessionContexts (which have cluster client caches) while the context is removed in parallel. The hope is that SessionContext instance should goes out of scope sooner than the double of the TTL.